### DISCUSSION TITLE
lnd v0.18.4-beta

### DISCUSSION THREAD

**Discussion by guggero** [lightningnetwork/lnd#9371] [Announcements]
This is a minor release which ships the features required for building custom channels, alongside the usual bug fixes and stability improvements.

Of note are the `bitcoind v28.0` compatibility fix and an optimization for payment requests (invoices) that removes unnecessary size expansion. See [release notes](https://github.com/lightningnetwork/lnd/blob/0-18-4-branch/docs/release-notes/release-notes-0.18.4.md) for full set of changes.

# Verifying the Release

In order to verify the release, you'll need to have `gpg` or `gpg2` installed on your system. Once you've obtained a copy (and hopefully verified that as well), you'll first need to import the keys that have signed this release if you haven't done so already: 

```
curl https://raw.githubusercontent.com/lightningnetwork/lnd/master/scripts/keys/roasbeef.asc | gpg --import
```

Once you have the required PGP keys, you can verify the release (assuming `manifest-roasbeef-v0.18.4-beta.sig` and `manifest-v0.18.4-beta.txt` are in the current directory) with:

```
gpg --verify manifest-roasbeef-v0.18.4-beta.sig manifest-v0.18.4-beta.txt
```

You should see the following if the verification was successful:

```
gpg: Signature made Wed Sep 30 17:35:20 2020 PDT
gpg:                using RSA key 60A1FA7DA5BFF08BDCBBE7903BBD59E99B280306
gpg: Good signature from "Olaoluwa Osuntokun <laolu32@gmail.com>" [ultimate]
```

That will verify the signature of the manifest file, which ensures integrity and authenticity of the archive you've downloaded locally containing the binaries. Next, depending on your operating system, you should then re-compute the `sha256` hash of the archive with `shasum -a 256 <filename>`, compare it with the corresponding one in the manifest file, and ensure they match *exactly*.

## Verifying the Release Timestamp 

From this new version onwards, in addition time-stamping the _git tag_ with [OpenTimestamps](https://opentimestamps.org/), we'll also now timestamp the manifest file along with its signature. Two new files are now included along with the rest of our release artifacts: ` manifest-roasbeef-v0.18.4-beta.txt.asc.ots`.

Assuming you have the opentimestamps client installed locally, the timestamps can be verified with the following commands: 
```
ots verify manifest-roasbeef-v0.18.4-beta.sig.ots -f manifest-roasbeef-v0.18.4-beta.sig
```

Alternatively, [the OpenTimestamps website](https://opentimestamps.org/) can be used to verify timestamps if one doesn't have a `bitcoind` instance accessible locally.

These timestamps should give users confidence in the integrity of this release even after the key that signed the release expires. 

## Verifying the Release Binaries

Our release binaries are fully reproducible. Third parties are able to verify that the release binaries were produced properly without having to trust the release manager(s). See our [reproducible builds guide](https://github.com/lightningnetwork/lnd/blob/master/docs/release.md) for how this can be achieved.
The release binaries are compiled with `go1.22.6`, which is required by verifiers to arrive at the same ones.
They include the following build tags: `autopilotrpc`, `signrpc`, `walletrpc`, `chainrpc`, `invoicesrpc`, `neutrinorpc`, `routerrpc`, `watchtowerrpc`, `monitoring`, `peersrpc`, `kvdb_postrgres`, `kvdb_etcd` and `kvdb_sqlite`. Note that these are already included in the release script, so they do not need to be provided.

The `make release` command can be used to ensure one rebuilds with all the same flags used for the release. If one wishes to build for only a single platform, then `make release sys=<OS-ARCH> tag=<tag>` can be used. 

Finally, you can also verify the _tag_ itself with the following command:

```
$ git verify-tag v0.18.4-beta
gpg: Signature made Tue Sep 15 18:55:00 2020 PDT
gpg:                using RSA key 60A1FA7DA5BFF08BDCBBE7903BBD59E99B280306
gpg: Good signature from "Olaoluwa Osuntokun <laolu32@gmail.com>" [ultimate]
```

## Verifying the Docker Images

To verify the `lnd` and `lncli` binaries inside the docker images against the signed, reproducible release binaries, there is a verification script in the image that can be called (before starting the container for example):

```shell
$ docker run --rm --entrypoint="" lightninglabs/lnd:v0.18.4-beta /verify-install.sh v0.18.4-beta
$ OK=$?
$ if [ "$OK" -ne "0" ]; then echo "Verification failed!"; exit 1; done
$ docker run lightninglabs/lnd [command-line options]
```

# Building the Contained Release 

Users are able to rebuild the target release themselves without having to fetch any of the dependencies. In order to do so, assuming
that `vendor.tar.gz` and `lnd-source-v0.18.4-beta.tar.gz` are in the current directory, follow these steps:

```
tar -xvzf vendor.tar.gz
tar -xvzf lnd-source-v0.18.4-beta.tar.gz
GO111MODULE=on go install -v -mod=vendor -ldflags "-X github.com/lightningnetwork/lnd/build.Commit=v0.18.4-beta" ./cmd/lnd
GO111MODULE=on go install -v -mod=vendor -ldflags "-X github.com/lightningnetwork/lnd/build.Commit=v0.18.4-beta" ./cmd/lncli
```

The `-mod=vendor` flag tells the `go build` command that it doesn't need to fetch the dependencies, and instead, they're all enclosed in the local vendor directory. 

Additionally, it's now possible to use the [enclosed `release.sh` script to bundle a release for a _specific_ system like so](https://github.com/lightningnetwork/lnd/pull/2191): 

```
make release sys="linux-arm64 darwin-amd64"
```

⚡️⚡️⚡️ OK, now to the rest of the release notes! ⚡️⚡️⚡️

# Release Notes

https://github.com/lightningnetwork/lnd/blob/0-18-4-branch/docs/release-notes/release-notes-0.18.4.md

# Contributors (Alphabetical Order)
* Boris Nagaev
* Elle Mouton
* ffranr
* George Tsagkarelis
* Jonathan Harvey-Buschel
* Keagan McClelland
* Matt Morehouse
* Olaoluwa Osuntokun
* Oliver Gugger
* Yong Yu
* Ziggie

<hr /><em>This discussion was created from the release <a href='https://github.com/lightningnetwork/lnd/releases/tag/v0.18.4-beta'>lnd v0.18.4-beta</a>.</em>

**Comment by schildbach** [comment]
Are validators aware there is one signature missing?

```
ERROR: Not enough valid signatures found!
  Valid signatures found: 4
  Valid signatures required: 5
```

**Reply by guggero** [reply]
There are 5, but Roasbeef's key changed, which isn't in the side branch that `v0.18.4-beta` was built from. If you run the script on `master`, you get 5 valid signatures.

**Reply by schildbach** [reply]
What does this mean for the Docker build? I'm currently doing

```
$ git switch --detach v0.18.4-beta

$ docker build --build-arg checkout=v0.18.4-beta --pull --no-cache -t mytag .

$ docker run --rm --entrypoint="" mytag /verify-install.sh v0.18.4-beta
```

Which of those do I need to switch to master?

I tried various combinations but get all sorts of errors, from hash mismatches to "ERROR: Invalid expected version detected: master".

**Reply by guggero** [reply]
Hmm, right... After building the image as outlined above, you can run this to verify using the keys from master:
```
git checkout master
docker run --rm --entrypoint="" -v $(pwd)/scripts/verify-install.sh:/verify-install.sh -v $(pwd)/scripts/keys/:/keys/ mytag /verify-install.sh v0.18.4-beta
```

**Reply by schildbach** [reply]
Thanks, that worked!

**Comment by twofaktor** [comment]


**Reply by guggero** [reply]
https://github.com/lightningnetwork/lnd/issues/9373
